### PR TITLE
Enable setting custom vertex ids in TP3 interface #147 

### LINKED ISSUE #147: Enable custom id setting in JanusGraphBlueprintsTransaction
JanusGraphFeatures suggests that the use of custom vertex ids through TinkerPop Structure [depends on the allowVertexIdSetting() value](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/JanusGraphFeatures.java#L179). However the TitanBlueprintsTransaction implementation [forbids custom ids regardless of the allowVertexIdSetting](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/JanusGraphBlueprintsTransaction.java#L102). I then unfortunately had to drill down to [vendor-specific implementation to use custom vertex ids](https://github.com/classmethod/tupl-titan-storage-backend/blob/titan10/src/test/java/jp/classmethod/titan/TuplStorageSetup.java#L100) even though they should be supported, despite [having enabled setting vertex IDs](https://github.com/classmethod/tupl-titan-storage-backend/blob/titan10/src/test/java/jp/classmethod/titan/TuplStorageSetup.java#L124).

### PR DESCRIPTION
**Pull request by amcp**
fixes #147, another bug fix that didn't make it into Titan by the end of 2015
Signed-off-by: Alexander Patrikalakis <amcp@amazon.co.jp>

### CODE REVIEW

**Comment by sjudeng** [comment]
I wonder if this might solve the issue I saw with custom vertex ids in upgrading to TinkerPop 3.2.3 (see notes in #78). To resolve there I updated test GraphComputerImplementations to include configuration around id generation but I wasn't happy with that approach.

**Comment by amcp** [comment]
@sjudeng I've got to put in for the night it being 2AM here, but if this interests you and given the extremely small size of the fix I encourage you to pick up from here and rip out the config workaround you put in GraphComputerImplementations to see if it fixes the issue you encountered.

**Comment by sjudeng** [comment]
I took a look at this branch. Initially I was having test failures of the form `The provided key/value array must have a String or T on even array indices`, which I resolved with the below update.

```patch
--- a/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/JanusGraphBlueprintsTransaction.java
+++ b/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/JanusGraphBlueprintsTransaction.java
@@ -115,7 +115,8 @@ public abstract class JanusGraphBlueprintsTransaction implements JanusGraphTrans
             label = (labelValue instanceof VertexLabel)?(VertexLabel)labelValue:getOrCreateVertexLabel((String) labelValue);
         }
 
-        final JanusGraphVertex vertex = addVertex(ElementHelper.getIdValue(keyValues).orElse(null), label);
+        final Number id = (Number) ElementHelper.getIdValue(keyValues).orElse(null);
+        JanusGraphVertex vertex = addVertex(id != null ? id.longValue() : null, label);
```

I then tried running through tests in the PeerPressureTest suite via InMemoryJanusGraphComputerTest, which use custom vertex ids. But these gave errors at [StandardJanusGraphTx.java#L514](https://github.com/amcp/janusgraph/blob/ac5172d6ace652854263f068105b387eefaf3cad/janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java#L514). I experimented with an additional conditional there for `graph.getConfiguration().allowVertexIdSetting() && UserVertex.is(vertexId)` but that also failed (`UserVertex.is(vertexId)` was false). I also tried forcing it past that error but then ended up with an error lower down at [IDManager.java#L474](https://github.com/amcp/janusgraph/blob/ac5172d6ace652854263f068105b387eefaf3cad/janusgraph-core/src/main/java/org/janusgraph/graphdb/idmanagement/IDManager.java#L474).

**Comment by sjudeng** [comment]
@amcp I rebased and added a commit to this PR. The commit includes a fix to JanusGraphBlueprintsTransaction to resolve test failures as described in the comment above. Also during testing I found that the existing documented method for converting user ids to vertex ids, `JanusGraphId#toVertexId(long)`, is no longer valid as it doesn't take into account partitioning bits used in JanusGraph vertex assignment. Attempts to flush vertices with custom ids created using this method can lead to assert/IllegalArgumentException errors. To resolve this issue I added methods to IDManager for converting user vertex ids to/from JanusGraph vertex ids and deprecated JanusGraphId.

All default and TinkerPop tests are passing with these updates. Given my involvement in the implementation I think it would be best to have another independent review.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/idmanagement/IDManager.java [hunk 1]
```diff
@@ -567,6 +567,33 @@ public long getPartitionedVertexId(long partitionedVertexId, long otherPartition
...
 
+    /**
+     * Converts a user provided long id into a JanusGraph vertex id. The id must be positive and less than {@link #getVertexCountBound()}.
+     * This method is useful when providing ids during vertex creation via {@link org.apache.tinkerpop.gremlin.structure.Graph#addVertex(Object...)}.
+     *
+     * @param id long id
+     * @return a corresponding JanusGraph vertex id
+     * @see #fromVertexId(long)
+     */
+    public long toVertexId(long id) {
```
Comment (amcp) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/idmanagement/IDManager.java]: can you test this method?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/idmanagement/IDManager.java [hunk 2]
```diff
@@ -567,6 +567,33 @@ public long getPartitionedVertexId(long partitionedVertexId, long otherPartition
...
+    }
+
+    /**
+     * Converts a JanusGraph vertex id to the user provided id as the inverse mapping of {@link #toVertexId(long)}.
+     *
+     * @param id JanusGraph vertex id (must be positive)
+     * @return original user provided id
+     * @see #toVertexId(long)
+     */
+    public long fromVertexId(long id) {
```
Comment (amcp) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/idmanagement/IDManager.java]: can you test this method?

**Comment by amcp** [comment]
@sjudeng I agree we need another review.
@analytically the content of this PR changed slightly so can you re-review please?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/idmanagement/IDManager.java (hunk 1 above)
Comment (sjudeng) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/idmanagement/IDManager.java]: @amcp There's a test added as part of the commit (see [VertexIDAssignerTest#testCustomIdAssignment](https://github.com/amcp/janusgraph/blob/fixBlueprintsCustomIds/janusgraph-test/src/test/java/org/janusgraph/graphdb/idmanagement/VertexIDAssignerTest.java#L186-L208)).

**Pull request merged**